File: sandbox/Damjan_Milosevic/backend/Services/LocationService.cs
using System.Globalization;
using Microsoft.EntityFrameworkCore;

public class LocationService
{
    private readonly AppDbContext _context;

    public LocationService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<List<Location>> GetLocationsByCategory(int category)
    {
        return await _context.Locations
            .Where(l => l.CategoryId == category)
            .ToListAsync();
    }

    public async Task<List<Location>> GetTopLocations(int category)
    {
        return await _context.Locations
            .Where(l => l.CategoryId == category)
            .OrderByDescending(l => l.Rating)
            .Take(3)
            .ToListAsync();
    }

    public async Task<List<Location>> GetByCity(string city)
    {
        return await _context.Locations
            .Where(l => l.City == city)
            .ToListAsync();
    }

    public async Task<List<Location>> Search(string name)
    {
        return await _context.Locations
            .Where(l => l.Name.Contains(name))
            .ToListAsync();
    }

    public async Task<Location> AddLocation(Location location)
    {
        _context.Locations.Add(location);
        await _context.SaveChangesAsync();
        return location;
    }

    public async Task<bool> DeleteLocation(int id)
    {
        var location = await _context.Locations.FindAsync(id);
        if (location == null)
        {
            return false;
        }

        _context.Locations.Remove(location);
        await _context.SaveChangesAsync();
        return true;
    }
}

File: sandbox/Damjan_Milosevic/frontend/src/app/services/location-service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../environments/environment.development';
import { Observable } from 'rxjs';
import { Location } from '../models/locations';

@Injectable({
  providedIn: 'root',
})
export class LocationService {

  

  constructor(private http:HttpClient){}
  
  getLocationsByCategory(category:number):Observable<Location[]>{
    return this.http.get<Location[]>(environment.apiUrl+`/location/category/${encodeURIComponent(category)}`);
  }

  getTopLocations(category:number):Observable<Location[]>{
    return this.http.get<Location[]>(environment.apiUrl+"/location/top/"+category);
  }

  getByCity(city:string):Observable<Location[]>{
    return this.http.get<Location[]>(environment.apiUrl+`/location/city/${encodeURIComponent(city)}`);
  }

  searchByName(name:string):Observable<Location[]>{
    return this.http.get<Location[]>(environment.apiUrl+`/location/search?name=${encodeURIComponent(name)}`);
  }

  addLocation(location: Location): Observable<Location> {
    return this.http.post<Location>(environment.apiUrl + '/location', location);
  }

  deleteLocation(location:number): Observable<void> {
    return this.http.delete<void>(environment.apiUrl + '/location/'+ location);
  }
}
